### PR TITLE
feat: mutate pods based on pod label

### PR DESCRIPTION
**Pull request by aramase**
Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

**Reason for Change**:

- Supports mutating based on `azure.workload.identity/use` label in the pod. This is done for backward compatibility.
- Adds warning as part of admission response if pod is missing the required label.

Screenshot 2022-11-30 at 11 26 36 AM

**Requirements**

- [ ] squashed commits
- [ ] included documentation
- [ ] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

part of #601 

**Please answer the following questions with yes/no**:

Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

- [ ] yes
- [ ] no

**Notes for Reviewers**:

### CODE REVIEW

File: test/e2e/webhook.go [hunk 1]
```diff
@@ -30,6 +30,7 @@ var _ = ginkgo.Describe("Webhook", func() {
 			[]string{"3600"},
 			nil,
 			nil,
+			nil,
 			false,
 		)
 		framework.ExpectNoError(err, "failed to create pod %s in %s", pod.Name, f.Namespace.Name)
```
Comment (enj) [inline test/e2e/webhook.go:36]: How difficult would it be to check for warnings here?

File: scripts/ci-e2e.sh [hunk 2]
```diff
@@ -94,9 +94,8 @@ test_helm_chart() {
     --debug \
     -v=5
   poll_webhook_readiness
-  # TODO(aramase: remove GINKGO_SKIP once helm chart is updated to use v0.12.0 to include the 
-  # new proxy probe feature.
-  GINKGO_SKIP=Proxy make test-e2e-run
+  # TODO(aramase): remove GINKGO_SKIP once helm chart is updated to use v0.15.0
+  GINKGO_SKIP=Proxy\|should.mutate.a.labeled.pod make test-e2e-run
```
[the remark this answers is not in the document]
Comment (aramase) [reply scripts/ci-e2e.sh:98]: We do! `test_helm_chart` performs an upgrade test: 

1. Install the released version and run tests

https://github.com/Azure/azure-workload-identity/blob/e00e0d315583c827dab98d993dd8243c626e1c97/scripts/ci-e2e.sh#L87-L99

2. Upgrade to the CI built version and run the tests again

https://github.com/Azure/azure-workload-identity/blob/e00e0d315583c827dab98d993dd8243c626e1c97/scripts/ci-e2e.sh#L101-L113

I'm skipping theses new tests from the released version only.

List of tests we run:
1. Deploy the yaml from `manifest_staging/deploy` and run e2e tests.
2. Install latest release, run e2e tests, upgrade to CI build with latest helm charts from `manifest_staging/charts` and run e2e tests again.

[the remark this answers is not in the document]
Comment (enj) [reply scripts/ci-e2e.sh:98]: oo excellent 👍

Thanks for the explanation!

File: pkg/webhook/webhook.go [hunk 3]
```diff
@@ -34,6 +34,11 @@ var (
 	podLabelMissingWarning = fmt.Sprintf(`pod missing label '%s: "true"'. This label will be required in a future release for the webhook to mutate pod. Please add the label to the pod.`, UseWorkloadIdentityLabel)
 )
 
+const (
+	// warningAnnotationKey is the annotation key used to store warning messages in audit annotations
+	warningAnnotationKey = "warning"
```
Comment (enj) [inline pkg/webhook/webhook.go:39]: Check the docs of the `AuditAnnotations` field - these keys have to be prefixed based on the webhook.

Comment (aramase) [reply pkg/webhook/webhook.go:39]: Ah I misread https://github.com/kubernetes/kubernetes/blob/17bf864c1fc20e4badc9a6e0659c5190310462e1/pkg/apis/admission/types.go#L135-L140 and thought it's added after the response is returned. Let me modify it.

Comment (aramase) [reply pkg/webhook/webhook.go]: done!

File: test/e2e/webhook.go (hunk 1 above)
Comment (enj) [reply test/e2e/webhook.go:36]: See if something like this works:

```diff
diff --git a/test/e2e/webhook.go b/test/e2e/webhook.go
index 4bc56d2..71b7830 100644
--- a/test/e2e/webhook.go
+++ b/test/e2e/webhook.go
@@ -3,23 +3,45 @@
 package e2e
 
 import (
+	"bytes"
 	"context"
 	"fmt"
+	"io"
 	"strings"
+	"sync"
 
 	"github.com/Azure/azure-workload-identity/pkg/webhook"
 
 	"github.com/onsi/ginkgo/v2"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
 	"k8s.io/kubernetes/test/e2e/framework"
 	"k8s.io/utils/pointer"
 )
 
+var _ io.Writer = &syncBuffer{}
+
+type syncBuffer struct {
+	lock sync.Mutex
+	buf  bytes.Buffer
+}
+
+func (s *syncBuffer) Write(p []byte) (int, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	return s.buf.Write(p)
+}
+
 var _ = ginkgo.Describe("Webhook", func() {
 	f := framework.NewDefaultFramework("webhook")
 
 	ginkgo.It("should mutate a pod with a labeled service account", func() {
+		buf := &syncBuffer{}
+		rest.SetDefaultWarningHandler(rest.NewWarningWriter(buf, rest.WarningWriterOptions{}))
+		defer rest.SetDefaultWarningHandler(rest.NoWarnings{})
+
 		serviceAccount := createServiceAccount(f.ClientSet, f.Namespace.Name, f.Namespace.Name+"-sa", map[string]string{webhook.UseWorkloadIdentityLabel: "true"}, nil)
 		pod, err := createPodWithServiceAccount(
 			f.ClientSet,
@@ -35,6 +57,7 @@ var _ = ginkgo.Describe("Webhook", func() {
 		)
 		framework.ExpectNoError(err, "failed to create pod %s in %s", pod.Name, f.Namespace.Name)
 		validateMutatedPod(f, pod, nil)
+		framework.ExpectEqual(buf.buf.String(), "the warning text we expect to see", "did not see the correct warnings")
 	})
 
 	ginkgo.It("should mutate the init containers within a pod", func() {
```

**Review by enj — APPROVED** [review]

**Pull request merged**